pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -209,6 +276,87 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
         }
     }
 
+    /**
+     * Acknowledge message(s) for an ongoing transaction.
+     * <p>
+     * It can be of {@link AckType#Individual} or {@link AckType#Cumulative}. Single messages acked by ongoing
+     * transaction will be put in pending_ack state and only marked as deleted after transaction is committed.
+     * <p>
+     * Only one transaction is allowed to do cumulative ack on a subscription at a given time.
+     * If a transaction do multiple cumulative ack, only the one with largest position determined by
+     * {@link PositionImpl#compareTo(PositionImpl)} will be kept as it cover all position smaller than it.
+     * <p>
+     * If an ongoing transaction cumulative acked a message and then try to ack single message which is
+     * smaller than that one it cumulative acked, it'll succeed.
+     * <p>
+     * If transaction is aborted all messages acked by it will be put back to pending state.
+     *
+     * @param txnID                  TransactionID of an ongoing transaction trying to sck message.
+     * @param positions              {@link Position}(s) it try to ack.
+     * @param ackType                {@link AckType}.
+     * @throws TransactionConflictException
+     * @throws IllegalArgumentException
+     */
+    public synchronized void acknowledgeMessage(TxnID txnID, List<Position> positions, AckType ackType) throws TransactionConflictException {
+        checkArgument(txnID != null, "TransactionID can not be null.");
+        if (AckType.Cumulative == ackType) {
+            // Check if another transaction is already using cumulative ack on this subscription.
+            if (this.txnID != null && this.txnID != txnID) {
+                String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                        txnID.getLeastSigBits() + " try to cumulative ack message while transaction:"
+                        + this.txnID.getMostSigBits() + this.txnID.getLeastSigBits()
+                        + " already cumulative acked messages.";
+                log.error(errorMsg);
+                throw new TransactionConflictException(errorMsg);
+            }
+
+            if (positions.size() != 1) {
+                String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                txnID.getLeastSigBits() + " invalid cumulative ack received with multiple message ids.";
+                log.error(errorMsg);
+                throw new IllegalArgumentException(errorMsg);
```

Comment:
should this be TransactionConflictException?

Reply:
I see TransactionConflictException as exception thrown when there's conflict, e.g. re-ack from same transaction or single ack within another  transaction's cumulative ack range. 
But this case is purely because passed in wrong argument and we didn't even start to ack anything, so I think IllegalArgumentException is more proper here.

Reply:
sgtm